x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/correlation/BucketCorrelationAggregationBuilder.java

```diff
@@ -0,0 +1,130 @@
+import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+import org.elasticsearch.xpack.core.ml.utils.NamedXContentObjectHelper;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+public class BucketCorrelationAggregationBuilder extends AbstractPipelineAggregationBuilder<BucketCorrelationAggregationBuilder> {
+
+    public static final ParseField NAME = new ParseField("bucket_correlation");
+    private static final ParseField FUNCTION = new ParseField("function");
+
+    @SuppressWarnings("unchecked")
+    public static final ConstructingObjectParser<BucketCorrelationAggregationBuilder, String> PARSER = new ConstructingObjectParser<>(
+        NAME.getPreferredName(),
+        false,
+        (args, context) -> new BucketCorrelationAggregationBuilder(
+            context,
+            (String)args[0],
+            (CorrelationFunction)args[1]
+        )
+    );
+    static {
+        PARSER.declareString(ConstructingObjectParser.constructorArg(), BUCKETS_PATH_FIELD);
+        PARSER.declareNamedObject(
+            ConstructingObjectParser.constructorArg(),
+            (p, c, n) -> p.namedObject(CorrelationFunction.class, n, null),
+            FUNCTION
+        );
+    }
+
+    private final CorrelationFunction correlationFunction;
+
+    public BucketCorrelationAggregationBuilder(
+        String name,
+        String bucketsPath,
+        CorrelationFunction correlationFunction
+    ) {
+        super(
+            name,
+            NAME.getPreferredName(),
+            new String[] {bucketsPath}
+        );
+        this.correlationFunction = correlationFunction;
+    }
+
+    public BucketCorrelationAggregationBuilder(StreamInput in) throws IOException {
+        super(in, NAME.getPreferredName());
+        this.correlationFunction = in.readNamedWriteable(CorrelationFunction.class);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME.getPreferredName();
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeNamedWriteable(correlationFunction);
+    }
+
+    @Override
+    protected PipelineAggregator createInternal(Map<String, Object> metadata) {
+        return new BucketCorrelationAggregator(name, correlationFunction, bucketsPaths[0], metadata);
+    }
+
+    @Override
+    protected boolean overrideBucketsPath() {
+        return true;
+    }
+
+    @Override
+    protected XContentBuilder internalXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.field(BUCKETS_PATH_FIELD.getPreferredName(), bucketsPaths[0]);
+        NamedXContentObjectHelper.writeNamedObject(builder, params, FUNCTION.getPreferredName(), correlationFunction);
+        return builder;
+    }
+
+    @Override
+    protected void validate(ValidationContext context) {
+
+        final String firstAgg = bucketsPaths[0].split("[>\\.]")[0];
```

Comment:
> BucketMetricsPipelineAggregationBuilder

This class supports things we don't want to support like user provided formats + gap policies. I could overwrite those methods to make sure prevent them from being set internally (it can be prevented in the parser). 

I will see if extending that class will work.

Reply:
IIRC, I did not inherit from that class because of an initial design choice. But that has since changed. So, hopefully, it works nicely

Reply:
@dimitris-athanasiou I think to do this, I would need to do a refactoring of `BucketMetricsPipelineAggregationBuilder` to either allow `format` and `gapPolicy` to be accessed directly in the sub-class, or that all accesses go through their respective getters.

Reply:
Could we not pass in the zero gap policy?

Reply:
I'd feel more comfortable merging with the similar code and performing a more mechanical refactoring in a follow up then pushing it into this.

Reply:
> Could we not pass in the zero gap policy?

I am not sure where we would do that. Also, when serializing in XContent the gap_policy is written. I suppose we call `gapPolicy(GapPolicy)` in the subclass ctor, but that felt...messy (accessing a method in a ctor).